### ISSUE TITLE
Analyzer strong warnings hard to parse

### ISSUE THREAD

**Issue by srawlins** [dart-lang/sdk#27356]
I've tried for a minimal recreation of some code that is not happy with the new 1.20-dev strong analyzer warnings. It's not super minimal, but here it is:

``` dart
import 'dart:async';

class Cat {
  static final Camel<BlueHeron> camel = (BlueHeron _) => 'Sam';

  Bunny x() {
    var puppy = new Puppy();
    var brownBear = new BrownBear(puppy);
    return new Bunny(new Whale<BlueHeron>(), brownBear, camel);
    // ERROR: Could not infer type parameter T, Heron must be of type BlueHeron.
    // ERROR: The argument type 'BrownBear<Heron>' cannot be assigned to the parameter type 'Bear<BlueHeron>'.
  }
}

class BrownBear<T> extends Bear<T> {
  final PuppyFn<T, dynamic> puppyFn;

  BrownBear(this.puppyFn) : super([]);

  Future<Null> fn(Context context) async {
    await puppyFn([context]);
  }
}

class Bear<T> {
  List<T> ts;

  Bear(List<T> ts) {
    this.ts = ts;
  }
}

abstract class PuppyFn<T, D> {
  Future<Iterable<T>> call(List<D> list);
}

class Puppy implements PuppyFn<Heron, Context> {
  @override
  Future<List<Heron>> call(List<Context> list) =>
      new Future.value(list.first.heron);
}

class Context {
  Heron get heron => new Heron();
}

class Whale<T> {}

class Bunny<T> {
  Whale whale;
  Bear<T> bear;
  Camel<T> camel;

  Bunny(Whale<T> whale, Bear<T> bear, Camel<T> camel) {
    this.whale = whale;
    this.bear = bear;
    this.camel = camel;
  }
}

typedef String Camel<T>(T item);

class Heron {}

class BlueHeron extends Heron {}
```

Analyzer says:

```
$ xcodebuild/ReleaseX64/dart-sdk/bin/dartanalyzer --strong e2.dart
Analyzing [e2.dart]...
[error] Could not infer type parameter T, Heron must be of type BlueHeron. (/Users/srawlins/code/dart-repo3/sdk/e2.dart, line 9, col 16)
[error] The argument type 'BrownBear<Heron>' cannot be assigned to the parameter type 'Bear<BlueHeron>'. (/Users/srawlins/code/dart-repo3/sdk/e2.dart, line 9, col 46)
2 errors found.
```

This might need to be split into 2 bugs, but these errors seem related.
1. `Could not infer type parameter T, Heron must be of type BlueHeron.`
   So... is there a problem? A Heron _is_ a BlueHeron. You apparently did infer it?
2. `The argument type 'BrownBear<Heron>' cannot be assigned to the parameter type 'Bear<BlueHeron>'`
   It's not hard to see what its saying, in general. But it's really hard to find why the analyzer thinks that `brownBear` is a `BrownBear<Heron>`. I think it's inferred because of what Puppy implements...
   
   Can the analyzer give more context on warnings like this one? I know that this is like... a big difficult area of compiler design (good error messages), but if we could just get a little closer.

This was of course so much more difficult in real life because each of these classes and abstract classes were declared in different files, so you have to switch between 7 different files just to get a handle on why the analyzer decided a type annotation evaluated to one thing rather than the other.

**Comment by srawlins** [comment]
CC @keertip @dgrove

**Comment by srawlins** [comment]
No triage?

**Comment by jmesserly** [comment]
@srawlins first off, thanks so much for filing this! That is a really excellent reproduction. The real world complexity in there serves as a great illustration.

The good news: we can and definitely will do better on error messages :)

The bad news: right now we're still working on finishing up the implementation, so error messages may be a bit rough for a bit. It's crucial we fix https://github.com/dart-lang/sdk/issues/25490 for example.

My suggestion with a "failed to infer" error like this is to try passing a generic type explicitly. In this case `new Bunny<Heron>` for example, that might lead to other errors so you can discover where the problem originates.

I'll definitely think on this, perhaps there's a way to do better here prior to #25490, but it might end up being a bit too much of a detour.

**Comment by jmesserly** [comment]
And perhaps in the meantime the error message should suggest something like "try passing a generic type argument explicitly"?

**Comment by bwilkerson** [comment]
That would be good, although the suggestion ought to go in the correction message.

**Comment by jmesserly** [comment]
this is now fixed.
